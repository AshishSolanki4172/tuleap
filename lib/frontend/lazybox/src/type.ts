/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { HTMLTemplateResult, TemplateResult } from "lit/html.js";
import type { GroupCollection, LazyboxItem } from "./items/GroupCollection";
import type { HTMLTemplateStringProcessor } from "./index";

export interface Lazybox {
    setDropdownContent: (groups: GroupCollection) => void;
    resetSelection: () => void;
    setSelection: (selection: ReadonlyArray<LazyboxItem>) => void;
    destroy: () => void;
}

export type LazyboxSearchFieldCallback = (query: string) => void;

export type LazyboxSelectionCallback = (selected_value: unknown | null) => void;

export type LazyboxTemplatingCallback = (
    html: typeof HTMLTemplateStringProcessor,
    item: LazyboxItem
) => HTMLTemplateResult;

export type LazyboxNewItemCallback = () => void;

export type LazyBoxWithNewItemButton = {
    readonly new_item_button_label: string;
    readonly new_item_callback: LazyboxNewItemCallback;
};

type LazyBoxWithoutNewItemButton = {
    readonly new_item_callback?: undefined;
};

export type LazyboxOptions = (LazyBoxWithoutNewItemButton | LazyBoxWithNewItemButton) & {
    readonly placeholder: string;
    readonly search_input_placeholder: string;
    readonly is_multiple: boolean;
    readonly templating_callback: LazyboxTemplatingCallback;
    readonly selection_callback: LazyboxSelectionCallback;
    readonly search_field_callback: LazyboxSearchFieldCallback;
};

export type RenderedItemMap = Map<string, RenderedItem>;

export interface RenderedItem {
    id: string;
    template: TemplateResult;
    value: unknown;
    is_disabled: boolean;
    is_selected: boolean;
    group_id: string;
    element: Element;
}

export interface LazyboxComponent {
    wrapper_element: HTMLElement;
    lazybox_element: Element;
    dropdown_element: HTMLElement;
    selection_element: HTMLElement;
    placeholder_element: Element;
    dropdown_list_element: HTMLElement;
    search_field_element: HTMLInputElement;
}

export interface LazyboxSelectionStateSingle {
    selected_item: RenderedItem;
    selected_value_element: DocumentFragment;
}

export interface LazyboxSelectionStateMultiple {
    selected_items: Map<string, RenderedItem>;
    selected_values_elements: Map<string, Element>;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}

export interface ManageSelection {
    processSelection(item: Element): void;
    hasSelection(): boolean;
    updateSelectionAfterDropdownContentChange(): void;
    clearSelection(): void;
    setSelection(selection: ReadonlyArray<RenderedItem>): void;
}
