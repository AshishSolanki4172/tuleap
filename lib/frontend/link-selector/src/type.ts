/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { TemplateResult } from "lit/html.js";
import type { LinkSelectorSearchFieldCallback } from "./events/SearchFieldEventCallbackHandler";

export interface LinkSelector {
    destroy: () => void;
}

export interface LinkSelectorOptions {
    readonly placeholder?: string;
    search_field_callback: LinkSelectorSearchFieldCallback;
}

export type RenderedItemMap = Map<string, RenderedItem>;

export interface RenderedItem {
    id: string;
    template: TemplateResult;
    value: string;
    is_disabled: boolean;
    is_selected: boolean;
    group_id: string;
    element: Element;
    target_option: HTMLOptionElement;
}

export interface LinkSelectorComponent {
    wrapper_element: HTMLElement;
    link_selector_element: Element;
    dropdown_element: HTMLElement;
    selection_element: HTMLElement;
    placeholder_element: Element;
    dropdown_list_element: HTMLElement;
    search_field_element: HTMLInputElement;
}

export interface LinkSelectorSelectionStateSingle {
    selected_item: RenderedItem;
    selected_value_element: DocumentFragment;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}
