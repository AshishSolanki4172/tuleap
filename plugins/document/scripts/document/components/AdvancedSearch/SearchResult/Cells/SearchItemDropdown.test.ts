/**
 * Copyright (c) Enalean, 2022 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { Wrapper } from "@vue/test-utils";
import { shallowMount } from "@vue/test-utils";
import SearchItemDropdown from "./SearchItemDropdown.vue";
import localVue from "../../../../helpers/local-vue";
import type { Item, ItemSearchResult } from "../../../../type";
import DropDownMenuTreeView from "../../../Folder/DropDown/DropDownMenuTreeView.vue";
import { createStoreMock } from "@tuleap/core/scripts/vue-components/store-wrapper-jest";

const observe = jest.fn();

window.ResizeObserver =
    window.ResizeObserver ||
    jest.fn().mockImplementation(() => ({
        observe,
        unobserve: jest.fn(),
    }));

describe("SearchItemDropdown", () => {
    let wrapper: Wrapper<SearchItemDropdown>;
    let $store = {
        dispatch: jest.fn(),
    };
    let parent_container: HTMLElement;
    let addEventListener: jest.SpyInstance;

    beforeEach(() => {
        $store = createStoreMock({});
        $store.dispatch.mockImplementation((action) => {
            if (action === "loadDocument") {
                return Promise.resolve({
                    id: 111,
                } as Item);
            }
        });

        const mount_point = document.createElement("div");

        parent_container = document.createElement("div");
        parent_container.classList.add("document-search-table-container");
        parent_container.appendChild(mount_point);
        addEventListener = jest.spyOn(parent_container, "addEventListener");

        document.body.appendChild(parent_container);

        wrapper = shallowMount(SearchItemDropdown, {
            localVue,
            propsData: {
                item: {
                    id: 111,
                } as ItemSearchResult,
            },
            mocks: {
                $store,
            },
            attachTo: mount_point,
        });
    });

    afterEach(() => {
        wrapper.destroy();
    });

    it("should render the dropdown with the menu detached", () => {
        expect(document.body).toMatchSnapshot();
    });

    it("should display a spinner if real item is not loaded", () => {
        expect(wrapper.find("[data-test=spinner]").exists()).toBe(true);
        expect(wrapper.findComponent(DropDownMenuTreeView).exists()).toBe(false);
    });

    it("should observe the resize of the body to adapt the position of trigger button", () => {
        expect(observe).toHaveBeenCalledWith(document.body);
    });

    it("should observe the scroll of the table container to adapt the position of trigger button", () => {
        expect(addEventListener).toHaveBeenCalledWith("scroll", expect.anything(), {
            passive: true,
        });
    });

    it("should display the menu as soon as the user click on the trigger and the real item is loaded", () => {
        return new Promise((done) => {
            wrapper.find("[data-test=trigger").trigger("click");

            process.nextTick(() => {
                expect(wrapper.find("[data-test=spinner]").exists()).toBe(false);
                expect(wrapper.findComponent(DropDownMenuTreeView).exists()).toBe(true);
                done(null);
            });
        });
    });
});
