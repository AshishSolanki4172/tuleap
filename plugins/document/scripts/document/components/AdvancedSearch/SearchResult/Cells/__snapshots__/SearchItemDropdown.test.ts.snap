// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchItemDropdown should render the dropdown with the menu detached 1`] = `
<body>
  <div>
    <div
      class="tlp-dropdown document-search-dropdown"
    >
      <button
        aria-label="Open dropdown menu"
        class="tlp-button-primary tlp-button-small"
        data-dropdown="trigger"
        data-test="trigger"
        id="document-search-dropdown-trigger-111"
        type="button"
      >
        <i
          class="fa fa-ellipsis-h"
        />
         
        <i
          class="fa fa-caret-down tlp-button-icon-right"
        />
      </button>
       
    </div>
  </div>
  <div
    class="tlp-dropdown-menu document-dropdown-menu tlp-dropdown-menu-right"
    data-dropdown="menu"
    data-triggered-by="document-search-dropdown-trigger-111"
    role="menu"
  >
    <div
      class="document-search-dropdown-spinner"
      data-test="spinner"
    >
      <i
        aria-hidden="true"
        class="fas fa-circle-notch fa-spin"
      />
    </div>
  </div>
</body>
`;
